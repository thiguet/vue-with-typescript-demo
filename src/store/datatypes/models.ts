export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product {
    name: string;
    measure: keyof typeof Measures;
    qtd: number;
    minQtd: number;
    img?: string;
}

export interface User {
    name: string;
    email: string;
}

export interface LoginSubmit {
    name: string;
    pass: string;
}

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
}
