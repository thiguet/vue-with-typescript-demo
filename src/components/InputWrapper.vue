<template>
    <fieldset>
        <legend>
            <span class="accessibility">{{ legend }}</span>
        </legend>
        <label class="label">
            <span>{{ fixedLabel }}</span>
        </label>
        <slot></slot>
    </fieldset>
</template>

<script lang="ts">
import Vue from 'vue';

import Component from 'vue-class-component';

import { Prop } from 'vue-property-decorator';

@Component({
    name: 'InputWrapper',
})
export default class InputWrapper extends Vue {
    @Prop({ required: true }) readonly label!: string;

    @Prop() readonly legend!: string;

    get fixedLabel() {
        return `${this.label}:`;
    }
}
</script>

<style scoped>
fieldset {
    display: flex;
    flex-direction: column;
    border: none;
    width: 100%;
    padding: 0;
    margin-bottom: 20px;
}

label {
    display: flex;
    flex-direction: column;
    font-size: 14px;
    margin-bottom: 5px;
    width: 100%;
}

label ~ input,
label ~ select {
    outline: none;
}

label:focus {
    border: 12px solid black;
}
</style>
