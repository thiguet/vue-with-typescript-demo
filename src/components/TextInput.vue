<template>
    <input
        :id="id"
        :type="type"
        :name="id"
        class="input"
        :value="value"
        @input="onInput"
    />
</template>

<script lang="ts">
import Vue from 'vue';

import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';

@Component({
    name: 'TextInput',
})
export default class TextInput extends Vue {
    @Prop({ required: true }) readonly id!: string;

    @Prop({ default: 'text' }) readonly type!: string;

    @Prop({ required: true }) readonly value!: string;

    @Prop({ required: true }) readonly setValue!: Function;

    private onInput(e: Event) {
        this.setValue((e.target as HTMLInputElement).value);
    }
}
</script>

<style scoped>
input {
    border: none;
    border-bottom: 1px solid var(--border-color);
    width: 100%;
    font-family: inherit;
    font-size: var(--font-input-size);
}
</style>
