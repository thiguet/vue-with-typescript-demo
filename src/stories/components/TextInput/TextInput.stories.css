@import url('../../app.css');

.input {
    border-bottom: 1px solid var(--border-color) !important;
}
