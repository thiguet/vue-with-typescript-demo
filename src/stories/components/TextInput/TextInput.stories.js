import TextInput from '@/components/TextInput.vue';
import './TextInput.stories.css';

export default {
    title: 'Example/Components/TextInput',
    component: TextInput,
};

const Template = (_args, { argTypes }) => ({
    props: Object.keys(argTypes),
    components: { TextInput },
    template: '<text-input v-bind="$props"/>',
});

let value = 'someVal';

const setValue = (val) => {
    value = val;
};

export const Default = Template.bind({});
Default.args = {
    id: 'some-id',
    value,
    setValue,
};
