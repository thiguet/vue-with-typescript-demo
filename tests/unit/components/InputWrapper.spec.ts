import { shallowMount } from '@vue/test-utils';
import TextInput from '@/components/TextInput.vue';
import InputWrapper from '@/components/InputWrapper.vue';
import faker from 'faker';
import { TextInputProps, InputWrapperProps } from './models';

describe('InputWrapper.vue', () => {
    let props: InputWrapperProps;

    const build = () => {
        const textInputWrapper = {
            render(h: Function) {
                const textInputData: TextInputProps = {
                    id: 'someid',
                    value: 'somevalue',
                    setValue: jest.fn(),
                };

                return h(TextInput, {
                    props: textInputData,
                });
            },
        };

        const wrapper = shallowMount(InputWrapper, {
            propsData: { ...props },
            slots: { default: textInputWrapper },
        });

        return {
            wrapper,
            input: () => wrapper.findComponent(TextInput),
            labelEl: () => wrapper.find('.label'),
            legend: () => wrapper.find('legend'),
        };
    };

    beforeEach(() => {
        props = {
            label: faker.random.words(),
            legend: faker.random.words(),
        };
    });

    it('renders component', () => {
        props.label = 'someFixedValidLabel';
        props.legend = 'someFixedValidLegend';

        const { wrapper } = build();

        expect(wrapper).toMatchSnapshot();
    });

    it('renders main child components', () => {
        const { labelEl, legend } = build();

        expect(legend().exists()).toBe(true);
        expect(labelEl().exists()).toBe(true);
    });

    it('passes props to the component', () => {
        const { wrapper } = build();

        const { label, legend } = wrapper.props();

        expect(label).toBe(props.label);
        expect(legend).toBe(props.legend);
    });

    it('sets propeties to label element', () => {
        const { wrapper, labelEl } = build();

        expect(labelEl().text()).toBe(wrapper.props().label + ':');
    });

    it('sets legends text', () => {
        const { wrapper, legend } = build();

        expect(legend().text()).toBe(wrapper.props().legend);
    });
});
