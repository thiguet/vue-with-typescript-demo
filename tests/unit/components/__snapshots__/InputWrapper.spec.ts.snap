// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputWrapper.vue renders component 1`] = `
<fieldset>
  <legend><span class="accessibility">someFixedValidLegend</span></legend> <label class="label"><span>someFixedValidLabel:</span></label>
  <textinput-stub id="someid" type="text" value="somevalue" setvalue="function mockConstructor() {
        return fn.apply(this, arguments);
      }"></textinput-stub>
</fieldset>
`;
